File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int prefix = 1;
        int suffix = 1;
        int product = 1;
        int[] output = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            output[i] = product * prefix;
            product *= nums[i];
        }
        product = 1;
        for (int i = nums.length - 1; i >=0 ; i--) {
            output[i] *= suffix * product;
            product *= nums[i];
        }

        return output;
    }
}

File: Data Structures & Algorithms/find-duplicate-integer/submission-1.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        index1 = 0
        index2 = 0

        while 1:
            index1 = nums[index1]
            index2 = nums[nums[index2]]
        
            if index1 == index2:
                break
        
        index1 = 0
        while index1 != index2:
            index1 = nums[index1]
            index2 = nums[index2]
        
        return index1
